js_modules/dagster-ui/packages/ui-core/src/overview/assets/AssetTimelineDataFetcher.tsx

```diff
@@ -0,0 +1,414 @@
+    return () => {
+      isCanceled = true;
+      clearInterval(interval);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [client, didCodeLocationNodesChange, didRangeMeaningfullyChange]);
+
+  const value = React.useMemo(() => {
+    return {
+      subscribeToAssetData: (key: AssetKeyInput, listener: Listener) => {
+        const stringKey = tokenForAssetKey(key);
+        assetSubscriptions.current[stringKey] = listener;
+        const data = runsCache.current.assets[stringKey];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete assetSubscriptions.current[stringKey];
+        };
+      },
+      subscribeToGroupData: (groupName: string, listener: Listener) => {
+        groupSubscriptions.current[groupName] = listener;
+        const data = runsCache.current.groups[groupName];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete groupSubscriptions.current[groupName];
+        };
+      },
+      subscribeToCodeLocationData: (codeLocation: string, listener: Listener) => {
+        codeLocationSubscriptions.current[codeLocation] = listener;
+        const data = runsCache.current.locations[codeLocation];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete codeLocationSubscriptions.current[codeLocation];
+        };
+      },
+    };
+  }, []);
+  return <DataFetcherContext.Provider value={value}>{children}</DataFetcherContext.Provider>;
+};
+
+function groupRunsByGroup(runs: RunWithAssetsFragment[], assetToGroup: Record<string, string>) {
+  const groupedRuns: Record<
+    string,
+    {
+      assets: Record<string, RunWithAssetsFragment[]>;
+      runs: Set<RunWithAssetsFragment>;
+    }
+  > = {};
+
+  runs.forEach((run) => {
+    const {assets} = run;
+    assets.forEach((asset) => {
+      const stringKey = tokenForAssetKey(asset.key);
+      const group = assetToGroup[stringKey]!;
+      groupedRuns[group] = groupedRuns[group] || {
+        assets: {},
+        runs: new Set(),
+      };
+      groupedRuns[group]!.runs.add(run);
+      groupedRuns[group]!.assets[stringKey] = groupedRuns[group]!.assets[stringKey] || [];
+      groupedRuns[group]!.assets[stringKey]!.push(run);
+    });
+  });
+
+  return groupedRuns;
+}
+
+export const ASSET_TIMELINE_QUERY = gql`
+  query AssetTimelineQuery(
+    $inProgressFilter: RunsFilter!
+    $terminatedFilter: RunsFilter!
+    $tickCursor: Float
+    $ticksUntil: Float
+  ) {
+    unterminated: runsOrError(filter: $inProgressFilter) {
+      ... on Runs {
+        results {
+          id
+          ...RunWithAssetsFragment
+        }
+      }
+      ...PythonErrorFragment
+    }
+    terminated: runsOrError(filter: $terminatedFilter) {
+      ... on Runs {
+        results {
+          id
+          ...RunWithAssetsFragment
+        }
+      }
+      ...PythonErrorFragment
+    }
+    workspaceOrError {
```

Comment:
I wonder if we could / should pull all this from the existing copy in WorkspaceContext?  Other parts of the app don't really poll for the workspace info. I think by the time this query is running on the page, this info has already been fetched for the workspace context